lib/protobuf/decoder.ex

```diff
@@ -196,14 +196,93 @@ defmodule Protobuf.Decoder do
     val = if map?, do: %{embedded_msg.key => embedded_msg.value}, else: embedded_msg
     val = if oneof, do: {name_atom, val}, else: val
 
-    cond do
-      is_nil(current) and repeated? -> [val]
-      is_nil(current) and not repeated? -> val
-      repeated? -> [val | current]
-      not repeated? -> Map.merge(current, val)
+    if repeated? do
+      # List.wrap/1 wraps nil into [].
+      [val | List.wrap(current)]
+    else
+      # If the field is embedded but not repeated, it means that we need to merge the existing
+      # embedded message together with the new embedded message. If there is not existing embedded
+      # message, we default to a new empty message.
+      current =
+        cond do
+          current -> current
+          oneof -> {name_atom, type.new!([])}
+          true -> type.new!([])
+        end
+
+      deep_merge(current, val, type.__message_props__())
     end
   end
 
+  # If the field is a oneof, we merge its value and keep the tag.
+  defp deep_merge(_oneof1 = {tag, val1}, _oneof2 = {tag, val2}, props) do
+    {tag, deep_merge(val1, val2, props)}
+  end
+
+  # If the two fields to merge are the same message, we merge it by merging their fields.
+  defp deep_merge(%mod{} = msg1, %mod{} = msg2, %MessageProps{} = props) do
+    merged =
+      Enum.reduce(props.field_props, msg2, fn {_number, field_prop}, acc ->
+        %FieldProps{name_atom: field_name, oneof: oneof} = field_prop
+
+        case {msg1, msg2} do
+          {%{^field_name => val1}, %{^field_name => val2}} ->
+            if oneof do
+              %{acc | field_name => val2}
```

Comment:
Above oneofs are merged but here we overwrite with the second value, what's the difference?

Reply:
That they are of a "different" type, that is, when we get here the second oneof is another value of that oneof, so they cannot be merged reasonably. 

```proto
oneof body {
  string a = 1;
  string b = 2;
}
```

What I mean is that if we're here we're trying to merge `a` with `b`, so `b` always wins regardless of the types. If you try to merge `a` with `a`, then we do the "deep merging" in case they're two repeated fields or nested messages.